airbyte-workers/src/main/java/io/airbyte/workers/helper/FailureHelper.java

```diff
@@ -34,18 +36,62 @@ public static FailureReason genericFailure(final Throwable t, final Long jobId,
         .withMetadata(jobAndAttemptMetadata(jobId, attemptNumber));
   }
 
+  public static FailureReason genericFailure(final AirbyteTraceMessage m, final Long jobId, final Integer attemptNumber) {
+    return new FailureReason()
+        .withInternalMessage(m.getError().getInternalMessage())
+        .withStacktrace(m.getError().getStackTrace())
+        .withTimestamp(m.getEmittedAt().longValue())
+        .withMetadata(traceMessageMetadata(jobId, attemptNumber));
+  }
+
   public static FailureReason sourceFailure(final Throwable t, final Long jobId, final Integer attemptNumber) {
     return genericFailure(t, jobId, attemptNumber)
         .withFailureOrigin(FailureOrigin.SOURCE)
         .withExternalMessage("Something went wrong within the source connector");
   }
 
+  public static FailureReason sourceFailure(final AirbyteTraceMessage m, final Long jobId, final Integer attemptNumber) {
+    return genericFailure(m, jobId, attemptNumber)
+        .withFailureOrigin(FailureOrigin.SOURCE)
+        .withExternalMessage(m.getError().getMessage());
+  }
+
   public static FailureReason destinationFailure(final Throwable t, final Long jobId, final Integer attemptNumber) {
     return genericFailure(t, jobId, attemptNumber)
         .withFailureOrigin(FailureOrigin.DESTINATION)
         .withExternalMessage("Something went wrong within the destination connector");
   }
 
+  public static FailureReason destinationFailure(final AirbyteTraceMessage m, final Long jobId, final Integer attemptNumber) {
+    return genericFailure(m, jobId, attemptNumber)
+        .withFailureOrigin(FailureOrigin.DESTINATION)
+        .withExternalMessage(m.getError().getMessage());
+  }
+
+  public static FailureReason errorTraceMessageFailure(final AirbyteTraceMessage sourceMessage,
+                                                       final AirbyteTraceMessage destinationMessage,
+                                                       final Long jobId,
+                                                       final Integer attempt) {
+    if (sourceMessage == null && destinationMessage == null) {
+      return null;
+    }
+
+    if (sourceMessage != null && destinationMessage == null) {
+      return sourceFailure(sourceMessage, jobId, attempt);
+    }
+
+    if (sourceMessage == null && destinationMessage != null) {
+      return destinationFailure(destinationMessage, jobId, attempt);
+    }
+
+    if (sourceMessage.getEmittedAt() < destinationMessage.getEmittedAt()) {
```

Comment:
Why do we have this logic here. Shouldn't we combine the messages instead?

Reply:
we want to choose the earliest trace message as the failure reason, so if both the source and destination emit failure messages, we need to compare the times they were emitted to choose the earliest one